### PR TITLE
Refactor ActivityVisitor interface

### PR DESCRIPTION
**Pull request by gow**
## What changed?
No functional change. Just refactoring some code to reduce unnecessary abstractions.

## Why?
To make the code a bit less complicated.

## How did you test it?
Existing tests. Also verified that the previous version and the new version had good coverage.

## Potential risks
N/A

## Documentation
N/A

## Is hotfix candidate?
No

### CODE REVIEW

**Review by ychebotarev — APPROVED** [review]

File: service/history/workflow/mutable_state_impl.go [hunk 1]
```diff
@@ -4825,12 +4825,42 @@ func (ms *MutableStateImpl) RetryActivity(
...
+		makeBackoffAlgorithm(delay),
+	)
+	if retryState != enumspb.RETRY_STATE_IN_PROGRESS {
+		return retryState, nil
+	}
+
+	nextScheduledTime := now.Add(retryBackoff)
 	nextAttempt := ai.Attempt + 1
-	if err := ms.taskGenerator.GenerateActivityRetryTasks(ai.ScheduledEventId, activityVisitor.NextScheduledTime(), nextAttempt); err != nil {
+	if err := ms.taskGenerator.GenerateActivityRetryTasks(ai.ScheduledEventId, nextScheduledTime, nextAttempt); err != nil {
```
Comment (yycptt) [inline service/history/workflow/mutable_state_impl.go:4863]: nit: I'd move this logic after updateActivityInfoForRetries and revert the changes previously made to GenerateActivityRetryTasks, so that GenerateActivityRetryTasks still takes in activityInfo and generates task based on that (instead of assuming things like activity version == mutable state current version).

**Review by yycptt — APPROVED** [review]

File: service/history/workflow/mutable_state_impl.go (hunk 1 above)
Comment (gow) [reply service/history/workflow/mutable_state_impl.go]: I made the changes. But there was a test case specifically checking that the sizes should not be updated when we cannot generate retry activity task. So I'm updating the sizes after the task has been generated. Please take another quick look.

**Review by yycptt — APPROVED** [review]
